segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/exts3/ExtS3Storage.java

```
@@ -0,0 +1,445 @@
+ *
+ * Each segment is represented as a single Object on the underlying storage.
+ *
+ * Approach to fencing:
+ *
+ * This implementation works under the assumption that data is only appended and never modified.
+ * Each block of data has an offset assigned to it and Pravega always writes the same data to the same offset.
+ *
+ * With this assumption the only flow when a write call is made to the same offset twice is when ownership of the
+ * segment changes from one host to another and both the hosts are writing to it.
+ *
+ * As PutObject calls to with the same start-offset to an Exts3 object are idempotent (any attempt to re-write data with the same file offset does not
+ * cause any form of inconsistency), fencing is not required.
+ *
+ *
+ * In the absence of locking this is the expected behavior in case of ownership change: both the hosts will keep
+ * writing the same data at the same offset till the time the earlier owner gets a notification that it is not the
+ * current owner. Once the earlier owner received this notification, it stops writing to the segment.
+ */
+
+@Slf4j
+public class ExtS3Storage implements Storage {
+
+    //region members
+
+    private final ExtS3StorageConfig config;
+    private final ExecutorService executor;
+    private S3Client client = null;
+
+    //endregion
+
+    //region constructor
+
+    public ExtS3Storage(ExtS3StorageConfig config, ExecutorService executor) {
+        Preconditions.checkNotNull(config, "config");
+        Preconditions.checkNotNull(executor, "executor");
+        this.config = config;
+        this.executor = executor;
+
+    }
+
+    //endregion
+
+
+    //region Storage implementation
+
+    /**
+     * Initialize is a no op here as we do not need a locking mechanism in case of file system write.
+     * @param containerEpoch The Container Epoch to initialize with (ignored here).
+     */
+    @Override
+    public void initialize(long containerEpoch) {
+        if (client == null) {
```

You need to throw an exception if already initialized